code/modules/atmospherics/machinery/datum_pipeline.dm

```diff
@@ -231,25 +232,28 @@
 			continue
 		gas_mixture_list += pipeline.other_airs
 		gas_mixture_list += pipeline.air
-		for(var/atmosmch in pipeline.other_atmosmch)
-			if (istype(atmosmch, /obj/machinery/atmospherics/components/binary/valve))
-				var/obj/machinery/atmospherics/components/binary/valve/considered_valve = atmosmch
-				if(considered_valve.on)
-					pipeline_list |= considered_valve.parents[1]
-					pipeline_list |= considered_valve.parents[2]
-			else if (istype(atmosmch, /obj/machinery/atmospherics/components/unary/portables_connector))
-				var/obj/machinery/atmospherics/components/unary/portables_connector/considered_connector = atmosmch
-				if(considered_connector.connected_device)
-					gas_mixture_list += considered_connector.connected_device.return_air()
+		for(var/obj/machinery/atmospherics/components/atmosmch as anything in pipeline.other_atmosmch)
+			if(QDELETED(atmosmch))
+				continue
+
+			var/new_pipelines = atmosmch.returnPipenetsForReconcilation(src)
+			if(new_pipelines)
+				pipeline_list |= new_pipelines
+
+			var/new_airs = atmosmch.returnAirsForReconcilation(src)
+			if(new_airs)
+				gas_mixture_list |= new_airs
```

Comment:
How expensive is this? have you profiled it? I'd like to keep this proc as trim as possible even if that does make it ugly, because it's a consistent load since the main loop is always going to be equalizing by nature

Reply:
```
Self   | Total  | Real   | Overtime | Calls  |
---------------------------------------------|
2.308  | 4.343  | 4.364  | 0.000    | 20838  | Before change
4.947  | 7.311  | 7.308  | 0.000    | 20743  | After change
2.577  | 4.937  | 4.941  | 0.000    | 20896  | After change + optimization
```

Reply:
❤️